sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,131 @@
+- Decide to support a smaller no. of backends for kubernetes officially
+
+
+## Apiserver spending too much cpu/memory in duplicate Endpoints serializations
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our scale tests).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible
+- Do better batching of EP updates in endpoints-controller: This is more like a bandaid, as the duplicate serializations would still happen, but we at least reduce the total no. of watch events (at the cost of increased latency of ep updates potentially). Also, this would not help too much with cases where pods of services are toggling between Ready and Failed states
+
+
+## Endpoints can’t be efficiently PATCH’ed due to its current API schema
+
+#### Issue
+
+There were some [improvements](https://github.com/kubernetes/kubernetes/issues/47597#issuecomment-311698194) around endpoints-controller, but one issue that still remains that we couldn’t make progress on was to allow useful patching for EP objects (i.e patch contains only the endpoint entry to be changed). Due to the way the object is defined in the API, we seem to be not able to make progress without any breaking API changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API. Some properties we may like to have for the new API:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Sharded into multiple etcd key/value pairs? (potentially in its own etcd)
+  - Stable sharding & ordering, with a ready property instead of separate lists
+  - Port ranges and other features we’ve put off for a long time
+
+
+## Endpoints updates quadratic in the no. of endpoints
+
+#### Issue
+
+This is the equivalent issue for GET as that of the above issue for PUT. Every time a new endpoint is added to the endpoints object, the whole endpoints object is returned by the apiserver to watch. In essence, if the no. of endpoints are N, the total no. of endpoints delivered to a watch are O(N^2) in the worst case. We want to avoid this. Also note that this behavior can lead to O(N^2) space consumption in etcd due to storing revisions of the endpoints object, each time a new endpoint is appended.
```

Comment:
I don't fully agree with the description here. This is quadratic even if we won't do any GETs and watches.

The problem basically is:
- we want to do rolling upgrade of a service backends
- whenever one pod changes, you update the Endpoints object and for that you update the object consisting of N endpoints
- since you do that N times, in total you write N^2 pieces of data (because each object is of size O(N))

If you add watches into the picture it's even worse, because there is watch coming from every node, so that is multiplied by the size of cluster..

Reply:
I've rephrased the points to make the above clearer, PTAL.